In [43]:
import numpy as np
import matplotlib.pyplot as plt
from sklearn.metrics import accuracy_score

import pandas as pd

data = pd.read_csv("/Users/efang/Desktop/coding/Intro-to-ML/CSDS340/data/pima-indians-diabetes.csv", header=None)


In [44]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, MinMaxScaler, Normalizer, PowerTransformer, RobustScaler, MaxAbsScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
from scipy import stats

X = data.iloc[:, :-1]
y = data.iloc[:, -1]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5, random_state=1)

scalers = {
    'StandardScaler': StandardScaler(),
    'MinMaxScaler': MinMaxScaler(),
    'Normalizer': Normalizer(),
    'PowerTransformer': PowerTransformer(),
    'RobustScaler': RobustScaler(),
    'MaxAbsScaler': MaxAbsScaler()
}


k_values = range(1, 20)


results = []


for scaler_name, scaler in scalers.items():
   
    X_train_scaled = scaler.fit_transform(X_train)
    X_test_scaled = scaler.transform(X_test)
    
    for k in k_values:
        knn = KNeighborsClassifier(n_neighbors=k, metric='manhattan')
        knn.fit(X_train_scaled, y_train)
        
   
        y_pred = knn.predict(X_test_scaled)
        accuracy = accuracy_score(y_test, y_pred)
        

        results.append({'Scaler': scaler_name, 'k': k, 'Accuracy': accuracy})


results_df = pd.DataFrame(results)


best_result = results_df.loc[results_df['Accuracy'].idxmax()]


print(best_result)


Scaler      StandardScaler
k                       15
Accuracy          0.776042
Name: 14, dtype: object


The highest accuracy I was able to obtain is 77.6%.